src/script_interface/particle_data/ParticleHandle.cpp

```diff
@@ -114,6 +115,78 @@ static auto get_gamma_safe(Variant const &value) {
 }
 #endif // THERMOSTAT_PER_PARTICLE
 
+static bool force_code_coverage(bool value) { return value; }
+
+/**
+ * @brief Check allowlist of valid propagation modes combinations.
+ * Note for developers: when enabling new propagation mode combinations,
+ * make sure every single line of this function has code coverage.
+ */
+static auto is_valid_propagation_combination(int propagation) {
+  assert(propagation >= 0);
+  // check allowlist
+  switch (propagation) {
+  // only one mode
+  case NONE: // NOLINT(bugprone-branch-clone)
+    return force_code_coverage(true);
+  case SYSTEM_DEFAULT:
+    return force_code_coverage(true);
+  case TRANS_NEWTON:
+    return force_code_coverage(true);
+  case TRANS_BROWNIAN:
+    return force_code_coverage(true);
+  case TRANS_LANGEVIN:
+    return force_code_coverage(true);
+#ifdef NPT
+  case TRANS_LANGEVIN_NPT:
+    return force_code_coverage(true);
+#endif
+#ifdef ROTATION
+  case ROT_EULER:
+    return force_code_coverage(true);
+  case ROT_BROWNIAN:
+    return force_code_coverage(true);
+  case ROT_LANGEVIN:
+    return force_code_coverage(true);
+#endif // ROTATION
+  case TRANS_LB_MOMENTUM_EXCHANGE:
+    return force_code_coverage(true);
+#ifdef VIRTUAL_SITES
+  case TRANS_LB_TRACER:
+    return force_code_coverage(true);
+#endif
+#ifdef ROTATION
+  // same mode for translation and rotation
+  case TRANS_NEWTON | ROT_EULER:
+    return force_code_coverage(true);
+  case TRANS_LANGEVIN | ROT_LANGEVIN:
+    return force_code_coverage(true);
+  case TRANS_BROWNIAN | ROT_BROWNIAN:
+    return force_code_coverage(true);
+#ifdef STOKESIAN_DYNAMICS
+  case TRANS_STOKESIAN | ROT_STOKESIAN:
+    return force_code_coverage(true);
+#endif
+    // other mode combinations
+#ifdef NPT
+  case TRANS_LANGEVIN_NPT | ROT_EULER:
+    return force_code_coverage(true);
```

Comment:
TRANS_LANGEVIN_NPT | ROT_LANGEVIN. ROT_EULER woudl mean positoins are thermalied but orientaoitns not, that would not make sense.

Reply:
Well, ESPResSo's NpT implementation never thermalized angular velocities... But you know how I feel about our NpT implementation anyway :)

```python
import espressomd
system = espressomd.System(box_l=[1., 1., 1.], time_step=0.01)
system.cell_system.skin = 0.
p = system.part.add(pos=[0, 0, 0], v=[1, 2, 3], omega_lab=[1, 2, 3], rotation=3 * [True])
system.thermostat.set_npt(kT=0., gamma0=2., gammav=1e-6, seed=42)
system.integrator.set_isotropic_npt(ext_pressure=0.01, piston=1e6)
system.integrator.run(1)
print(f"{p.v=}")
print(f"{p.omega_lab=}")
```
Output on the python branch and in releases 4.2.1 and 4.1.4:
```
p.v=array([0.9801, 1.9602, 2.9403])
p.omega_lab=array([1., 2., 3.])
```

Reply:
addressed by #4843